iree/base/dynamic_library_test.cc

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/base/dynamic_library.h"
+
+#include <string>
+
+#include "iree/base/dynamic_library_test_library_embed.h"
+#include "iree/base/file_io.h"
+#include "iree/base/status.h"
+#include "iree/base/status_matchers.h"
+#include "iree/base/target_platform.h"
+#include "iree/testing/gtest.h"
+
+namespace iree {
+namespace {
+
+static const char* kUnknownName = "library_that_does_not_exist.so";
+
+class DynamicLibraryTest : public ::testing::Test {
+ public:
+  static void SetUpTestCase() {
+    std::string base_name = "dynamic_library_test_library";
+    ASSERT_OK_AND_ASSIGN(library_temp_path_, file_io::GetTempFile(base_name));
+#if defined(IREE_PLATFORM_WINDOWS)
+    library_temp_path_ += ".dll";
+#else
+    library_temp_path_ += ".so";
+#endif
+
+    const auto* file_toc = dynamic_library_test_library_create();
+    absl::string_view file_data(reinterpret_cast<const char*>(file_toc->data),
+                                file_toc->size);
+    ASSERT_OK(file_io::SetFileContents(library_temp_path_, file_data));
+
+    LOG(INFO) << "Embedded test library written to temp path: "
+              << library_temp_path_;
+  }
+
+  static std::string library_temp_path_;
+};
+
+std::string DynamicLibraryTest::library_temp_path_;
```

Comment:
Isn't this only necessary if you give it a default value?

Reply:
Nope, still required here:

```
dynamic_library_test.obj : error LNK2019: unresolved external symbol "public: static class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > iree::`anonymous namespace'::DynamicLibraryTest::library_temp_path_" (?library_temp_path_@DynamicLibraryTest@?A0x8ad6c487@iree@@2V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@A) referenced in function "public: static void __cdecl iree::`anonymous namespace'::DynamicLibraryTest::SetUpTestCase(void)" (?SetUpTestCase@DynamicLibraryTest@?A0x8ad6c487@iree@@SAXXZ)
bazel-out\x64_windows-opt\bin\iree\base\dynamic_library_test.exe : fatal error LNK1120: 1 unresolved externals
```

Reply:
I hate C++